README.md
# YoutubeTranscriptSummarizer
A Chrome Extension which makes request to a backend REST API where it will perform NLP and respond with a summarized version of a YouTube transcript.

The project is divided into 2 parts:
1. ChromeExtension- A Chrome Extension which sends request to the backend.
2. youtubetranscriptsummarizer- server which process the request and sends it back to the client as a HTTP response.

### Installation:
1. Clone the repo.
2. Go in the youtubetranscriptsummarizer folder and create, activate a virtual environment and install the requirements
   with pip or pip3 install -r requirements.txt .
3. Make a account on [ngrok.com](https://ngrok.com/) and run the following cell with the auth token from [dashboard](https://dashboard.ngrok.com/get-started/setup)
    ```
    ! ngrok authtoken YOUR_AUTH_TOKEN
    from pyngrok import ngrok
    ngrok.kill()
    #LOCALHOST_PORT can be seen by first running the run and seeing the port in localhost  
    ngrok.connect(LOCALHOST_PORT)
    ```
4. Everything is done, run the app now. You can acess the website using the url provided above(change http to https in front of the url).
5. Also change the server link in [popup.js](/ChromeExtension/popup.js) file in get request to API.
6. Next load the chrome extension folder [ChromeExtension](/ChromeExtension) unpacked from [chrome://extensions/](chrome://extensions/).
   Go to [youtube](https://youtube.com/) and open a video and click on summarize and see the summary.
7. connect to ngrok each time app is run a new time.

### Improvements:
1. Presently, only basic setup is done. Exception handling is left, if the url is not correct. 
   This would not happen when using chrome extension but if used independently then it can occur.
   Some expections: The video is live, url is not correct.
2. Summarization needs to be improved, time stamps can be added.
3. The transcript is presently working only for youtube videos having captions on, the audio and text processing are left.
4. There can be extra features of choosing language and length of summary. There can be option to convert the text 
   to audio . 
5. The summary of some 10-15 minutes as time stamps can be given, then a person can also search for a keyword, when it
   is discussed and can go directly to that time 
   
   
![](images/1.png)

![](images/2.png)
 


ChromeExtention/changes.txt
change font color with themẹ.
summary with time stamps.


ChromeExtention/contentScript.js
console.log("Content running");

chrome.runtime.onMessage.addListener(result =>{

  if(result.action == "print_summary"){  
    let sum = document.getElementById('summary');
      sum.innerHTML = result.summary;
      sum.style.textAlign = "left";
  }

  if(result.action == "load_summary"){
    console.log("trying to create div");
    let outer = document.getElementById('info');
        inner = document.createElement("div");
        inner.className = 'style-scope ytd-watch-flexy summary_box';
        inner.id = 'summary-box';
        let inner_text = document.createElement("P");
        inner_text.id = 'summary';
        inner.appendChild(inner_text);
        outer.appendChild(inner);
    let sum = document.getElementById('summary');
    sum.innerHTML = "Loading Summary.......";
    console.log("loading summary");
    sum.style.textAlign = "center";
}    

if(result.action == "clear"){
  let inner = document.getElementById('summary-box');
  if(inner != null)
      inner.remove();
}

});


youtubetranscriptsummarizer/app.py
from flask import Flask, render_template,request, make_response, jsonify, abort
from youtube_transcript_api import YouTubeTranscriptApi
from transformers import pipeline
from flask_cors import CORS
 
app = Flask(__name__)
CORS(app)


def get_trans(video_id):
    transcript = YouTubeTranscriptApi.get_transcript(video_id, languages=(['en']))
    data=''
    for i in transcript:
        data+=i['text']
        data+=' '
    return(data)


def summarize(id_video):
    data = get_trans(id_video)
    summarizer = pipeline('summarization')
    num_iter = int(len(data)/1000)
    summarized_text = []
    for i in range(0, num_iter + 1):
        start = 0
        start = i*1000
        end = (i+1)*1000
        out = summarizer(data[start:end], min_length = 10, max_length = 50)
        out = out[0]
        out = out['summary_text']
        summarized_text.append(out)
    summary = ' '.join(summarized_text)
    return(summary)


def retrieve_summary(url):
    video_id = url.split("=")[1]
    return(summarize(video_id))


@app.route('/')
def hello():
    return(render_template("index.html"))


@app.route('/api/summarize', methods=['GET'])
def index(): 
    url = request.args.get('youtube_url', '')
    transcript = retrieve_summary(url)
    if len(transcript)==0:
        abort(404)
    return transcript


@app.errorhandler(404)
def not_found(self):
    return make_response(jsonify({'error':'Not found'}),404)


if __name__== '__main__':
    app.run(debug=True)







youtubetranscriptsummarizer/app copy.py
from flask import Flask, render_template,request, make_response, jsonify, abort
from urllib.parse import urlparse, parse_qs
from youtube_transcript_api import YouTubeTranscriptApi
from transformers import pipeline
#from transformers import T5ForConditionalGeneration, T5Tokenizer
from flask_cors import CORS
 
app = Flask(__name__)
CORS(app)

def get_trans(video_id):  #cookie issue
    transcript = YouTubeTranscriptApi.get_transcript(video_id, languages=(['en']))
    data=''
    for i in transcript:
        data+=i['text']
        data+=' '
        print(len(data))
    return(data)

def summarize(video_id):
    data = get_trans(video_id)
    summarizer = pipeline('summarization')
    num_iter = int(len(data)/1000)
    summarized_text = []
    for i in range(0, num_iter + 1):
        start = 0
        start = i*1000
        end = (i+1)*1000
        out = summarizer(data[start:end], min_length = 10, max_length = 50)
        out = out[0]
        out = out['summary_text']
        summarized_text.append(out)
    summary = ' '.join(summarized_text)
    #summary = summarization(data)[0]['summary_text']
    #model = T5ForConditionalGeneration.from_pretrained('t5-base')
    #tokenizer = T5Tokenizer.from_pretrained("t5-base")
    #input = tokenizer.encode("summarize: " + data, return_tensors="pt", max_length= 1024, truncation=True)
    #summary=model.generate(
        #input,
        #max_length=150,
        #min_length=40
        #length_penalty=2.0,
        #num_beams=4,
        #early_stopping=True
    #)
    #return(tokenizer.decode(summary[0]))
    print(len(summary))
    return(summary)


#def get_audio(url):
#    from youtube_dl import YoutubeDL
 #   ydl_opts = {
  #      'format': 'bestaudio/best',
   #     'postprocessors': [{
    #        'key': 'FFmpegExtractAudio',
     #       'preferredcodec': 'mp3'
      #  }],
    #}
    #with YoutubeDL(ydl_opts) as ydl:
       #ydl.download([url])

def extract_video_id(url):
    query = urlparse(url)
    if query.hostname == 'youtu.be': return query.path[1:]
    if query.hostname in {'www.youtube.com', 'youtube.com'}:
        if query.path == '/watch': return parse_qs(query.query)['v'][0]
        if query.path[:7] == '/watch/': return query.path.split('/')[1]
        if query.path[:7] == '/embed/': return query.path.split('/')[2]
        if query.path[:3] == '/v/': return query.path.split('/')[2]
        # below is optional for playlists
        if query.path[:9] == '/playlist': return parse_qs(query.query)['list'][0]
       # returns None for invalid YouTube url

def retrieve_summary(url):
    youtube_api=  extract_video_id(url)
    return(summarize(youtube_api))

@app.route('/')
def hello():
    return(render_template("index.html"))

@app.route('/api/summarize', methods=['GET'])
def index(): 
    url = request.args.get('youtube_url', '')
    #url = " "
    transcript = retrieve_summary(url)
    if len(transcript)==0:
        abort(404)
    return transcript



@app.errorhandler(404)
def not_found(self):
    return make_response(jsonify({'error':'Not found'}),404)

if __name__== '__main__':
    app.run(debug=True)





